core/server/src/main/java/alluxio/master/file/meta/InodeDirectory.java

```diff
@@ -76,15 +77,20 @@ protected InodeDirectory getThis() {
    * @param child the inode to add
    */
   public void addChild(Inode<?> child) {
-    mChildren.add(child);
+    synchronized (child) {
```

Comment:
users should not need to worry about synchronization. Originally, IndexedSet handled synchronization across indexes.

If we allow users to also use index directly, we need to ensure that indexes are easy to use, and are thread-safe. What is the thread-safety of `FieldIndex`?